### PR TITLE
fix(developer): prevent buffer overrun in `u16tok`

### PR DESCRIPTION
**Pull request by mcdurdin**
Relates-to: #11814
See-also: #11894

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmcmplib/src/kmx_u16.cpp [hunk 1]
```diff
@@ -243,7 +243,7 @@ KMX_WCHAR * u16tok(KMX_WCHAR *p, const KMX_WCHAR ch,  KMX_WCHAR **ctx) {
   else {
     *ctx = NULL;
   }
-  return p;
+  return *p ? p : NULL;
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/kmx_u16.cpp:246]: Prevents delimiters at end of string from returning an empty string. (Matching `strtok` behaviour)

File: developer/src/kmcmplib/src/kmx_u16.cpp [hunk 2]
```diff
@@ -259,13 +259,13 @@ KMX_WCHAR * u16tok(KMX_WCHAR* p, const KMX_WCHAR* delim, KMX_WCHAR** ctx) {
 	if (*q) {
 		*q = 0;
 		q++;
-		while (u16chr(delim, *q)) q++;
+		while (*q && u16chr(delim, *q)) q++;
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/kmx_u16.cpp:262]: Prevent buffer overrun

**Comment by mcdurdin** [comment]
I've added a few more tests to #11894 which builds on this.

File: developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#include <gtest/gtest.h>
+#include "../src/kmx_u16.h"
+
+/*class kmx_u16_Test : public testing::Test {
+    protected:
+    	kmx_u16_Test() {}
+	    ~kmx_u16_Test() override {}
+	    void SetUp() override {}
+	    void TearDown() override {}
+};*/
```
Comment (darcywong00) [inline developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp:10]: Are these stubs needed?

**Review by jahorton — APPROVED** [review]

**Comment by markcsinclair** [comment]
I merged #11910 into #11894, so I'm waiting for #11910 to be merged into master before I submit #11894 for review.  Josh has approved #11910, so I'm hoping you will merge it into master soon.

File: developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp (hunk 3 above)
Comment (mcdurdin) [reply developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp:10]: Not really, but we may need them again soon as we extend the tests.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.73-alpha](https://keyman.com/downloads/releases/alpha/18.0.73)

**Pull request merged**